File: data_source_service/test/unitaries/controllers/note_controller_test.rb
require "test_helper"

class NoteControllerTest < ActionDispatch::IntegrationTest
  test "Test If Method \"get_all_notes\" Renders All User's Notes From Database" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_ID, UserConstants::USER_TO_BE_CREATED_USERNAME

    NoteUseCases.create_note NoteConstants::FIRST_NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.create_note NoteConstants::SECOND_NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    get NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s

    serialize_response_body = JSON.parse response.body

    assert_response :ok
    assert_equal NoteConstants::NUMBER_OF_NOTES_CREATED_BY_UNIQUE_USER, serialize_response_body.length
  end

  test "Test If Method \"get_all_notes\" Renders Not Found Status Code When User Does Not Exist" do
    get NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s

    assert_response :not_found
  end

  test "Test If Method \"get_note\" Renders Note From Database" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_ID, UserConstants::USER_TO_BE_CREATED_USERNAME

    NoteUseCases.create_note NoteConstants::NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    get "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}"

    serialize_response_body = JSON.parse response.body

    assert_response :ok
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_ID, serialize_response_body[NoteConstants::NOTE_MAP_ID_KEY]
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_TITLE, serialize_response_body[NoteConstants::NOTE_MAP_TITLE_KEY]
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_BODY, serialize_response_body[NoteConstants::NOTE_MAP_BODY_KEY]
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP, serialize_response_body[NoteConstants::NOTE_MAP_CREATED_AT_KEY]
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP, serialize_response_body[NoteConstants::NOTE_MAP_UPDATED_AT_KEY]
    assert_equal UserConstants::USER_TO_BE_CREATED_ID, serialize_response_body[NoteConstants::NOTE_MAP_USER_ID_KEY]
  end

  test "Test If Method \"get_note\" Renders Not Found Status Code When Note Does Not Exist" do
    get "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}"

    assert_response :not_found
  end

  test "Test If Method \"create_note\" Renders Created Status Code When Note Is Created On Database" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_USERNAME, UserConstants::USER_TO_BE_CREATED_ID

    post NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s, params: {
      id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      createdAt: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updatedAt: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP
    }.to_json

    created_note = NoteUseCases.get_note NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_response :created
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_ID, created_note.id
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_TITLE, created_note.title
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_BODY, created_note.body
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP, created_note.created_at
    assert_equal NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP, created_note.updated_at
    assert_equal UserConstants::USER_TO_BE_CREATED_ID, created_note.user_id
  end

  test "Test If Method \"create_user\" Renders Not Found Status Code When User Does Not Exist" do
    post NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s, params: {
      id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      createdAt: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updatedAt: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP
    }.to_json

    assert_response :not_found
  end

  test "Test If Method \"create_user\" Renders Conflict Status Code When There Is A Note With The Same Id Already Created" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_USERNAME, UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.create_note NoteConstants::NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    post NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s, params: {
      id: NoteConstants::NOTE_TO_BE_CREATED_ID,
      title: NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      body: NoteConstants::NOTE_TO_BE_CREATED_BODY,
      createdAt: NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      updatedAt: NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP
    }.to_json

    assert_response :conflict
  end

  test "Test If Method \"update_note\" Renders Ok Status Code When Note Is Updated On Database" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_USERNAME, UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.create_note NoteConstants::NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    patch "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}", params: {
      title: NoteConstants::UPDATED_NOTE_TITLE,
      body: NoteConstants::UPDATED_NOTE_BODY,
      updatedAt: NoteConstants::NOTE_TO_BE_UPDATED_UPDATED_AT_TIMESTAMP
    }.to_json

    updated_note = NoteUseCases.get_note NoteConstants::NOTE_TO_BE_CREATED_ID

    assert_response :ok
    assert_equal NoteConstants::UPDATED_NOTE_TITLE, updated_note.title
    assert_equal NoteConstants::UPDATED_NOTE_BODY, updated_note.body
    assert_equal NoteConstants::NOTE_TO_BE_UPDATED_UPDATED_AT_TIMESTAMP, updated_note.updated_at
  end

  test "Test If Method \"update_note\" Renders Not Found Status Code When Note Or User Do Not Exist" do
    patch "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}", params: {
      title: NoteConstants::UPDATED_NOTE_TITLE,
      body: NoteConstants::UPDATED_NOTE_BODY,
      updatedAt: NoteConstants::NOTE_TO_BE_UPDATED_UPDATED_AT_TIMESTAMP
    }.to_json

    assert_response :not_found
  end

  test "Test If Method \"delete_note\" Renders No Content Status Code When Note Is Deleted On Database" do
    UserUseCases.create_user UserConstants::USER_TO_BE_CREATED_USERNAME, UserConstants::USER_TO_BE_CREATED_ID

    NoteUseCases.create_note NoteConstants::NOTE_TO_BE_CREATED_ID,
      NoteConstants::NOTE_TO_BE_CREATED_TITLE,
      NoteConstants::NOTE_TO_BE_CREATED_BODY,
      NoteConstants::NOTE_TO_BE_CREATED_CREATED_AT_TIMESTAMP,
      NoteConstants::NOTE_TO_BE_CREATED_UPDATED_AT_TIMESTAMP,
      UserConstants::USER_TO_BE_CREATED_ID

    delete "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}"

    assert_response :no_content

    assert_raises ActiveRecord::RecordNotFound do
      NoteUseCases.get_note NoteConstants::NOTE_TO_BE_CREATED_ID
    end
  end

  test "Test If Method \"delete_note\" Renders Not Found Status Code When Note Or User Do Not Exist" do
    delete "#{NoteConstants::NOTE_ROUTE + UserConstants::USER_TO_BE_CREATED_ID.to_s}/#{NoteConstants::NOTE_TO_BE_CREATED_ID}"

    assert_response :not_found
  end
end

File: in_memory_cache_service/src/tests/unitaries/controllers/note_controller.test.ts
import {
  describe,
  beforeAll,
  beforeEach,
  afterAll,
  test,
  expect,
} from "@jest/globals";
import request from "supertest";
import { StatusCodes } from "http-status-codes";
import { app } from "../../../server/server";
import { NOTES_INDEX_ROUTE } from "../../../constants/routes_constants";
import {
  NOT_EXISTING_NOTE_ERROR_MESSAGE,
  NOT_EXISTING_NOTE_ERROR_NAME,
  NOTE_BODY,
  NOTE_CREATED_AT,
  NOTE_ID,
  NOTE_TITLE,
  NOTE_UPDATED_AT,
} from "../../../constants/domains/note_constants";
import {
  USER_ID,
  USER_USERNAME,
} from "../../../constants/domains/user_constants";
import UserRepository from "../../../domains/user/user_repository";
import NoteRepository from "../../../domains/note/note_repository";
import UserRepositoryFactory from "../../../domains/user/user_repository_factory";
import NoteRepositoryFactory from "../../../domains/note/note_repository_factory";
import UserEntityFactory from "../../../domains/user/user_entity_factory";
import NoteEntityFactory from "../../../domains/note/note_entity_factory";
import TestAgent from "supertest/lib/agent";

describe("Test Class NoteController Behavior", () => {
  let userRepository: UserRepository;
  let noteRepository: NoteRepository;
  let requestOnApp: TestAgent;

  beforeAll(async () => {
    userRepository = await UserRepositoryFactory.getInstance();
    noteRepository = await NoteRepositoryFactory.getInstance();

    requestOnApp = request(app);
  });

  beforeEach(async () => {
    try {
      await userRepository.deleteUser(USER_ID);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-empty
    } catch (error) {}
  });

  afterAll(async () => {
    try {
      await userRepository.deleteUser(USER_ID);
      // eslint-disable-next-line @typescript-eslint/no-unused-vars, no-empty
    } catch (error) {}
  });

  test('Test If Method "getAllNotes" Responds All Notes From User', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await userRepository.createUser(user);
    await noteRepository.createNote(note);

    const response = await requestOnApp.get(`${NOTES_INDEX_ROUTE}${USER_ID}/`);

    expect(response.status).toEqual(StatusCodes.OK);
    expect(response.body[0].id).toEqual(NOTE_ID);
    expect(response.body[0].title).toEqual(NOTE_TITLE);
    expect(response.body[0].body).toEqual(NOTE_BODY);
    expect(response.body[0].createdAt).toEqual(NOTE_CREATED_AT);
    expect(response.body[0].updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(response.body[0].userId).toEqual(USER_ID);
  });

  test('Test If Method "getAllNotes" Responds Not Found If User Does Not Exist', async () => {
    const response = await requestOnApp.get(`${NOTES_INDEX_ROUTE}${USER_ID}/`);

    expect(response.status).toEqual(StatusCodes.NOT_FOUND);
  });

  test('Test If Method "getNote" Responds Note From User', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await userRepository.createUser(user);
    await noteRepository.createNote(note);

    const response = await requestOnApp.get(
      `${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`,
    );

    expect(response.status).toEqual(StatusCodes.OK);
    expect(response.body.id).toEqual(NOTE_ID);
    expect(response.body.title).toEqual(NOTE_TITLE);
    expect(response.body.body).toEqual(NOTE_BODY);
    expect(response.body.createdAt).toEqual(NOTE_CREATED_AT);
    expect(response.body.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(response.body.userId).toEqual(USER_ID);
  });

  test('Test If Method "getNote" Responds Not Found If User Or Note Does Not Exist', async () => {
    const response = await requestOnApp.get(
      `${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`,
    );

    expect(response.status).toEqual(StatusCodes.NOT_FOUND);
  });

  test('Test If Method "createNote" Creates Note On In-Memory Database', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);

    await userRepository.createUser(user);

    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    const response = await requestOnApp
      .post(`${NOTES_INDEX_ROUTE}${USER_ID}/`)
      .send(note);

    const retrievedNote = await noteRepository.getNote(NOTE_ID);

    expect(response.status).toEqual(StatusCodes.CREATED);
    expect(retrievedNote.id).toEqual(NOTE_ID);
    expect(retrievedNote.title).toEqual(NOTE_TITLE);
    expect(retrievedNote.body).toEqual(NOTE_BODY);
    expect(retrievedNote.createdAt).toEqual(NOTE_CREATED_AT);
    expect(retrievedNote.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(retrievedNote.userId).toEqual(USER_ID);
  });

  test('Test If Method "createNote" Responds Not Found If User Does Not Exist', async () => {
    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    const response = await requestOnApp
      .post(`${NOTES_INDEX_ROUTE}${USER_ID}/`)
      .send(note);

    expect(response.status).toEqual(StatusCodes.NOT_FOUND);
  });

  test('Test If Method "createNote" Responds Conflict If Note Already Exists', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);

    await userRepository.createUser(user);

    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteRepository.createNote(note);

    const response = await requestOnApp
      .post(`${NOTES_INDEX_ROUTE}${USER_ID}/`)
      .send(note);

    expect(response.status).toEqual(StatusCodes.CONFLICT);
  });

  test('Test If Method "updateNote" Updates Note On In-Memory Database', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);

    await userRepository.createUser(user);

    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      "",
      "",
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteRepository.createNote(note);

    const updatedNote = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    const response = await requestOnApp
      .patch(`${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`)
      .send(updatedNote);

    expect(response.status).toEqual(StatusCodes.OK);

    const retrievedNote = await noteRepository.getNote(NOTE_ID);

    expect(retrievedNote.id).toEqual(NOTE_ID);
    expect(retrievedNote.title).toEqual(NOTE_TITLE);
    expect(retrievedNote.body).toEqual(NOTE_BODY);
    expect(retrievedNote.createdAt).toEqual(NOTE_CREATED_AT);
    expect(retrievedNote.updatedAt).toEqual(NOTE_UPDATED_AT);
    expect(retrievedNote.userId).toEqual(USER_ID);
  });

  test('Test If Method "updateNote" Responds Not Found If User Or Note Does Not Exist', async () => {
    const response = await requestOnApp.patch(
      `${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`,
    );

    expect(response.status).toEqual(StatusCodes.NOT_FOUND);
  });

  test('Test If Method "deleteNote" Deletes Note On In-Memory Database', async () => {
    const user = UserEntityFactory.getInstance(USER_ID, USER_USERNAME);

    await userRepository.createUser(user);

    const note = NoteEntityFactory.getInstance(
      NOTE_ID,
      NOTE_TITLE,
      NOTE_BODY,
      NOTE_CREATED_AT,
      NOTE_UPDATED_AT,
      USER_ID,
    );

    await noteRepository.createNote(note);

    const response = await requestOnApp.delete(
      `${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`,
    );

    expect(response.status).toEqual(StatusCodes.OK);

    try {
      await noteRepository.getNote(NOTE_ID);
    } catch (error) {
      const castedError = error as Error;

      expect(castedError.name).toEqual(NOT_EXISTING_NOTE_ERROR_NAME);
      expect(castedError.message).toEqual(NOT_EXISTING_NOTE_ERROR_MESSAGE);
    }
  });

  test('Test If Method "deleteNote" Responds Not Found If User Or Note Does Not Exist', async () => {
    const response = await requestOnApp.delete(
      `${NOTES_INDEX_ROUTE}${USER_ID}/${NOTE_ID}`,
    );

    expect(response.status).toEqual(StatusCodes.NOT_FOUND);
  });
});
